### DISCUSSION TITLE
Make constants autoload by default?

### DISCUSSION THREAD

**Discussion by jgaskins** [opal/opal#2209] [General]
I've been thinking a lot about load-time performance. It takes Chrome nearly 300ms on my machine to get to first render on my laptop — which isn't the worst thing, but I'm thinking more about Android devices, which have abysmal single-core performance.

My proposed solution is to make classes lazy-load by default. For example, [in this gist](https://gist.github.com/jgaskins/efb7f0894d0bd5f21853c721c369afb0), `bar.rb` would be part of the app bundle because it was loaded with a `require` directive, but that module would never be executed because the constant `Bar` was never used. No time would be spent on the overhead of defining that constant or defining/stubbing its methods.

I don't mean using introspection here on filenames or anything like what Rails does. The compiler can track where each expression that defines a constant exists. This can be used to wrap the compiled JS inside a function and declare all top-level constants defined within it (all `module X`, `class Y`, and `Z =`, etc) to point to that function — similar to what `require` does now, but adding another layer of laziness inside of that. Each top-level constant name would have an entry in some lookup table pointing to an array of those functions. and `const_missing` would look that up and execute them JS.

This might not save a _lot_ of startup time (it seems like most of the corelib constants are used on load), but it would save _some_ time and it would likely shine in large apps, cutting out a lot of classes/modules that never get used on an app's first render.

**Comment by meh** [comment]
What about `#const_get` and `#const_set`?

**Comment by jgaskins** [comment]
`const_get` invokes `const_missing` which would trigger this lazy load, but `const_set` definitely throws a wrench in the works. I don't think there's a way we could determine statically whether that's a top-level constant (unless it literally calls `Object.const_set`).

**Comment by jgaskins** [comment]
The only potential solutions I can think of are to have lazy constants as a compiler flag or to disable lazy constants automatically if `const_set` is used. Neither one is a particularly great solution because it means your app startup performance can depend entirely on whether a gem uses `const_set`.

**Comment by elia** [comment]
@jgaskins @meh maybe a mid-ground solution is to use `autoload` for some of the rarely used stuff from corelib, e.g. I don't know if I ever used Complex and only seldom use most of the stuff listed in [opal/opal.rb](https://github.com/opal/opal/blob/7ef3aa8a51f2fac3b8d55125a53e59b237b0c83e/opal/opal.rb#L4-L16):

```rb
require 'corelib/string/inheritance'
require 'corelib/string/encoding'
require 'corelib/math'
require 'corelib/complex'
require 'corelib/rational'
require 'corelib/time'
require 'corelib/struct'
require 'corelib/io'
require 'corelib/main'
require 'corelib/dir'
require 'corelib/file'
require 'corelib/process'
require 'corelib/random'
```

That's basically reiterating the opal/mini opal/base etc. solution but making it automatic. I'm aware that's not viable for everything in the corelib but stuff like `Math`, `Process`, `Dir` or `File` can be easily moved to `autoload`.

**Comment by jgaskins** [comment]
Indeed, I imagine that for things like Array, Hash, String, Number, and a lot of other stuff used by everything, autoload would be pointless indirection. It would probably only be useful in making writing the class boilerplate more uniform with the rest of the codebase.

**Comment by janbiedermann** [comment]
ruby-hyperloop implements constant autoloading, so we don't have to `require` everything always in order:
https://github.com/ruby-hyperloop/hyperloop-config/blob/edge/lib/hyperloop/autoloader.rb
https://github.com/ruby-hyperloop/hyperloop-config/blob/edge/lib/hyperloop/autoloader_starter.rb

except for the compiler to compile things, which is automatic, require_tree.
